### PR TITLE
tools/objectstore: check for wrong coll open_collection

### PR DESCRIPTION
**Pull request by pereman2**
Add missing error checks on open_collection. Before it caused too many seg faults on null collections.

before:
```bash
[pdiazbou@o08 build]$ bin/ceph-objectstore-tool --data-path ../build/dev/osd1 --op list-pgs
1.0s0
[pdiazbou@o08 build]$ 
[pdiazbou@o08 build]$ bin/ceph-objectstore-tool --data-path ../build/dev/osd1 --op list | grep benchmark_data_o08_2468610_object299
["1.0s0",{"oid":"benchmark_data_o08_2468610_object299","key":"","snapid":-2,"hash":1448370175,"max":0,"pool":1,"namespace":"","shard_id":0,"max":0}]
[pdiazbou@o08 build]$ ugdb -- bin/ceph-objectstore-tool --data-path ../build/dev/osd1 --pgid 1.0 benchmark_data_o08_2468610_object299 list-omap
[pdiazbou@o08 build]$ bin/ceph-objectstore-tool --data-path ../build/dev/osd1 --pgid 1.0 benchmark_data_o08_2468610_object299 list-omap
*** Caught signal (Segmentation fault) **
 in thread 7fa38e432680 thread_name:ceph-objectstor
 ceph version 19.0.0-4613-g6e42fb9e3b4 (6e42fb9e3b4f8bee2901c3cf8a3f0c98b887cb8a) squid (dev)
 1: /lib64/libc.so.6(+0x3e6f0) [0x7fa38de3e6f0]
 2: (BlueStore::collection_list(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, ghobject_t const&, ghobject_t const&, int, std::vector<ghobject_t, std::allocator<ghobject_t> >*, ghobject_t*)+0x45) [0x5636fb02c551]
 3: (_action_on_all_objects_in_pg(ObjectStore*, coll_t, action_on_object_t&, bool)+0x438) [0x5636fa992760]
 4: (action_on_all_objects_in_exact_pg(ObjectStore*, coll_t, action_on_object_t&, bool)+0x62) [0x5636fa99393c]
 5: main()
 6: /lib64/libc.so.6(+0x29590) [0x7fa38de29590]
 7: __libc_start_main()
 8: _start()
Segmentation fault (core dumped)
```

after:
```bash
[pdiazbou@ build]$ bin/ceph-objectstore-tool --data-path ../build/dev/osd1 --pgid 1.0 benchmark_data_o08_2468610_object299 list-omap
Collection 1.0_head does not exist
```

Fixes: https://tracker.ceph.com/issues/66802
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/tools/ceph_objectstore_tool.cc [hunk 1]
```diff
@@ -1036,6 +1056,9 @@ int get_inc_osdmap(ObjectStore *store, epoch_t e, bufferlist& bl)
   if (store->read(ch,
 		  OSD::get_inc_osdmap_pobject_name(e),
 		  0, 0, bl) < 0) {
+  if (!ch) {
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:1059]: I think should go up, above the `read()`.

File: src/tools/ceph_objectstore_tool.cc [hunk 2]
```diff
@@ -1560,7 +1595,7 @@ int ObjectStoreTool::get_object(ObjectStore *store,
 
     spg_t coll_pgid;
     if (coll.is_pg(&coll_pgid) == false) {
-      cerr << "INTERNAL ERROR: Bad collection during import" << std::endl;
+      cerr << "INTERNAL ERROR: Bad collection during import: " << coll_pgid << " does not exist" << std::endl;
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:1598]: nit: new line?

File: src/tools/ceph_objectstore_tool.cc [hunk 3]
```diff
@@ -3206,7 +3340,7 @@ int dup(string srcpath, ObjectStore *src, string dstpath, ObjectStore *dst)
       if (ls.empty()) {
 	break;
       }
-      
+
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:3343]: :+1:

**Review by rzarzynski — COMMENTED** [review]
Do we want to backport? Is there a tacker for this fix? Asking b/c the commit description lacks the `Fixes` tag.

**Comment by pereman2** [comment]
I think we do @rzarzynski . Added tracker.

File: src/tools/ceph_objectstore_tool.cc [hunk 4]
```diff
@@ -1360,6 +1389,10 @@ int get_attrs(
   auto ch = store->open_collection(coll);
   if (debug)
     cerr << "\tattrs: len " << as.data.size() << std::endl;
+  if (!ch) {
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:1392]: nit: is putting the `ch` check after the `debug` one really intentional?

File: src/tools/ceph_objectstore_tool.cc [hunk 5]
```diff
@@ -2482,12 +2536,20 @@ int do_set_bytes(ObjectStore *store, coll_t coll,
   auto ch = store->open_collection(coll);
   if (!dry_run)
     store->queue_transaction(ch, std::move(*t));
+  if (!ch) {
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:2539]: Oops, too late for a non-dry run.

**Review by rzarzynski — COMMENTED** [review]
Missed these last time, sorry.

File: src/tools/ceph_objectstore_tool.cc (hunk 5 above)
Comment (pereman2) [reply src/tools/ceph_objectstore_tool.cc:2537]: oops

File: src/tools/ceph_objectstore_tool.cc [hunk 6]
```diff
@@ -2829,6 +2931,10 @@ int corrupt_info(ObjectStore *store, coll_t coll, ghobject_t &ghobj, Formatter*
     encode(oi, attr, -1);  /* fixme: using full features */
     t.setattr(coll, ghobj, OI_ATTR, attr);
     auto ch = store->open_collection(coll);
+    if (!ch) {
+    cerr << "Collection " << coll << " does not exist" << std::endl;
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:2935]: nit: messed indent.

**Review by rzarzynski — APPROVED** [review]

**Comment by pereman2** [comment]
jenkins test make check

**Pull request merged**